File: c/src/intro_sort.c
#include "../sorting_algorithms.h"
#include <stdlib.h>
#include <math.h>

static void swap(int *a, int *b) { int t = *a; *a = *b; *b = t; }

static size_t partition(int *arr, size_t low, size_t high) {
    int pivot = arr[high];
    size_t i = low;
    for (size_t j = low; j < high; ++j) {
        if (arr[j] < pivot) { swap(&arr[i], &arr[j]); ++i; }
    }
    swap(&arr[i], &arr[high]);
    return i;
}

static void heapify(int *arr, size_t n, size_t i, size_t offset) {
    size_t largest = i;
    size_t l = 2 * i + 1;
    size_t r = 2 * i + 2;
    if (l < n && arr[offset + l] > arr[offset + largest]) largest = l;
    if (r < n && arr[offset + r] > arr[offset + largest]) largest = r;
    if (largest != i) { swap(&arr[offset + i], &arr[offset + largest]); heapify(arr, n, largest, offset); }
}

static void heapsort_range(int *arr, size_t low, size_t high) {
    size_t n = high - low + 1;
    for (int i = (int)n / 2 - 1; i >= 0; --i) heapify(arr, n, (size_t)i, low);
    for (int i = (int)n - 1; i > 0; --i) { swap(&arr[low], &arr[low + (size_t)i]); heapify(arr, (size_t)i, 0, low); }
}

static void introsort_rec(int *arr, size_t low, size_t high, int depth_limit) {
    size_t size = high - low + 1;
    if (size <= 16) {
        for (size_t i = low + 1; i <= high; ++i) {
            int key = arr[i];
            size_t j = i;
            while (j > low && arr[j - 1] > key) { arr[j] = arr[j - 1]; --j; }
            arr[j] = key;
        }
        return;
    }
    if (depth_limit == 0) {
        heapsort_range(arr, low, high);
        return;
    }
    size_t p = partition(arr, low, high);
    if (p > low) introsort_rec(arr, low, p - 1, depth_limit - 1);
    if (p + 1 < high) introsort_rec(arr, p + 1, high, depth_limit - 1);
}

void intro_sort(int *arr, size_t n) {
    if (n <= 1) return;
    int depth_limit = (int)(2 * log((double)n));
    introsort_rec(arr, 0, n - 1, depth_limit);
}

File: python/sorting_algorithms/algos/introsort.py
import math
from typing import Sequence, Optional
from ..base import _setup_tracer, _finalize, SortTracer, Number


def introsort(data: Sequence[Number], *, trace: bool = False, tracer: Optional[SortTracer] = None):
    arr = list(data)
    tracer, ret_trace = _setup_tracer("introsort", trace, tracer)
    max_depth = int(math.log2(len(arr) + 1)) * 2 if arr else 0

    def _heapify(n: int, i: int):
        largest = i
        l, r = 2 * i + 1, 2 * i + 2
        if l < n and arr[l] > arr[largest]:
            largest = l
        if r < n and arr[r] > arr[largest]:
            largest = r
        if largest != i:
            arr[i], arr[largest] = arr[largest], arr[i]
            if tracer:
                tracer.record(arr, (i, largest), "heapify swap")
            _heapify(n, largest)

    def _heapsort(n: int):
        for i in range(n // 2 - 1, -1, -1):
            _heapify(n, i)
        for i in range(n - 1, 0, -1):
            arr[0], arr[i] = arr[i], arr[0]
            if tracer:
                tracer.record(arr, (0, i), "extract max")
            _heapify(i, 0)

    def _partition(lo: int, hi: int) -> int:
        pivot = arr[hi]
        i = lo
        for j in range(lo, hi):
            if tracer:
                tracer.record(arr, (j, hi), "compare")
            if arr[j] <= pivot:
                arr[i], arr[j] = arr[j], arr[i]
                if tracer:
                    tracer.record(arr, (i, j), "swap")
                i += 1
        arr[i], arr[hi] = arr[hi], arr[i]
        if tracer:
            tracer.record(arr, (i,), "pivot set")
        return i

    def _introsort(lo: int, hi: int, depth: int):
        size = hi - lo + 1
        if size <= 16:
            for i in range(lo + 1, hi + 1):
                key = arr[i]
                j = i - 1
                while j >= lo and arr[j] > key:
                    arr[j + 1] = arr[j]
                    j -= 1
                arr[j + 1] = key
                if tracer:
                    tracer.record(arr, (j + 1,), "insert")
            return
        if depth == 0:
            _heapsort(hi + 1)
            return
        p = _partition(lo, hi)
        _introsort(lo, p - 1, depth - 1)
        _introsort(p + 1, hi, depth - 1)

    if arr:
        _introsort(0, len(arr) - 1, max_depth)
    return _finalize(arr, tracer, ret_trace)
